### DISCUSSION TITLE
Battery Temp from an EG4 18K and EG4 PP not reporting correctly

### DISCUSSION THREAD

**Discussion by ruablack2** [celsworth/lxp-bridge#226] [General]
Hey thanks for making this awesome plugin. It's been a lifesaver with the terrible software/app on the EG4/Lux inverters. I'm using HA to do advanced scheduling of AC charge and battery discharge on a TOU plan.

Anyways, with the EG4 combo, I'm not getting the correct battery temp reported. It constantly reports 6C. I think this is maybe an issue with the EG4 specific firmware or the powerpro battery bms. On the inverter LCD it's self it only list he max and min cell temp and no generic battery temp. Now the bridge is pulling and logging the max and min temp. Now how would I go about getting that into a sensor in HA? I know enough around this code to be dangerous but this got a little above my head.

This is my print out of inputs 3,
`[2023-12-17T13:46:32.357 INFO lxp_bridge::mqtt] publishing: lxp/(serialnumber)/inputs/3 = {"max_chg_curr":20.0,"max_dischg_curr":20.0,"charge_volt_ref":56.0,"dischg_cut_volt":45.0,"bat_status_0":0,"bat_status_1":0,"bat_status_2":0,"bat_status_3":0,"bat_status_4":0,"bat_status_5":0,"bat_status_6":0,"bat_status_7":0,"bat_status_8":0,"bat_status_9":0,"bat_status_inv":3,"bat_count":1,"bat_capacity":280,"bat_current":2.85,"bms_event_1":0,"bms_event_2":0,"max_cell_voltage":34.05,"min_cell_voltage":33.94,"max_cell_temp":0.8,"min_cell_temp":0.5,"bms_fw_update_state":0,"cycle_count":35,"vbat_inv":54.3,"time":1702845992,"datalog":"(serialnumber)"}`

Now that max and min cell temp is correct. The inverter LCD and battery LCD read out 8C and 5C. The 6C that it being reported from inputs 2 is not correct.

`lxp/(serialnumber)/inputs/2 = {"e_pv_all":32.5,"e_pv_all_1":32.4,"e_pv_all_2":0.1,"e_pv_all_3":0.0,"e_inv_all":165.0,"e_rec_all":185.7,"e_chg_all":200.8,"e_dischg_all":158.9,"e_eps_all":1.1,"e_to_grid_all":50.7,"e_to_user_all":1354.0,"fault_code":0,"warning_code":0,"t_inner":25,"t_rad_1":41,"t_rad_2":35,"t_bat":6,"runtime":1563325,"time":1702845998,"datalog":"(serialnumber)"}`

<img width="720" alt="image" src="https://github.com/celsworth/lxp-bridge/assets/8856726/d953bcc1-f8a7-47ca-9511-6bec16e4feb3">

Also any change is being able to change the AC and PV (Battery Priority) charge power?

**Comment by celsworth** [comment]
Ah, nice to see it works at all with an EG4 inverter, I think you're the first to confirm it.

I can see a few things here, some of which I can fix:

The `max_cell_voltage` (and min) are clearly an order of magnitude out, 34.05 should be 3.405 - my inverter doesn't actually report these figures so it was always a guess, I can get that fixed.

Similarly `max_cell_temp` of 0.8 should be 8 and 0.5 should be 5, so similar deal but the other way. I'll fix that too.

`t_bat` is actually a bit different. I've not checked the EG4 documentation but on my Luxpower inverter this is only used when you have an NTC sensor plugged in (there's a dedicated port for it) and it's used more for monitoring lead acid batteries which can't report the temperatures on their own. So unless I have something plugged into that port, I get a meaningless figure out of it. I'm going to guess that the EG4 is the same (if it even exposes the NTC port at all). So I think this is actually expected, and you should just ignore `t_bat`. I'll check the EG4 protocol doc tomorrow to confirm, and see what it says.

I'll look into getting the charge power exposed to HA as well, shouldn't be too difficult, just to track down the right registers :)

**Reply by ruablack2** [reply]
Yeah same with `max_charge_current` and `max_discharge_current` they are off. Shows 20 instead of 200. 

Ahh, that makes sense about `t_batt`. Yeah I don't have a NTC hooked as the BMS is just reporting it's temps via CAN.

But is adding `max_cell_temp` and `min-cell-temp` something you have to add to the code or is it something I can add to a config file for the bridge to push that to mqtt?

**Reply by celsworth** [reply]
Hm the charge current ones is a bit confusing, because that's been correct in the past for Luxpower inverters. The EG4 documentation shows this is in units of 0.01A and I'm doing the correct `/100` for these. Can you double check that one, as either there's some very odd going on there or EG4's implementation doesn't agree with their own documentation :(

For your last question, they're already on MQTT, do you rather mean getting HA to see them as sensors? Yes, that is something I can add as well. I'll add it to the list :)

So max/min cell temp and the previously mentioned charge power settings?

**Reply by celsworth** [reply]
I've made a couple of merges with the above changes, if you're able to run the `dev` version (there's an entry in HA for it), you should start seeing some of the above changes. Take a copy of config, uninstall the stable addon, install dev addon, copy config back, should be good to go.

The changes will end up in a stable version sooner or later though, this is just for testing them if you want to.

**Reply by Quattrohead** [reply]
I am beta testing the EG4 18k version of this inverter and EG4 do not like external monitoring of the wireless dongle by HA or SA.
They are strongly suggesting that we monitor via the RS485 port  instead.
Personally, my HA server is next to the inverter so I can easily add a USB to RS485 right into a HA USB port.
Can you look into this as I fear Lux/EG4 will try to kill data access via wireless in the future.
